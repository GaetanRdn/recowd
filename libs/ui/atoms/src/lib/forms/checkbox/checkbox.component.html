<rc-material-icon *ngIf="!checked">check_box_outline_blank</rc-material-icon>
<rc-material-icon *ngIf="checked">check_box</rc-material-icon>

<label>
  <input
    #input
    [checked]="checked"
    [disabled]="disabled"
    [value]="value"
    type="checkbox"
  />
  <ng-content></ng-content>
</label>
