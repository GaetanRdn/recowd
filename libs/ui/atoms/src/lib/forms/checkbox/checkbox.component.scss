@use '../../assets/utils';

@mixin rc-checkbox($config) {
  $typography: utils.map-get-deep($config, typography, input);
  $colors: map-get($config, colors);
  $primary: utils.map-get-deep($config, colors, primary);

  rc-checkbox {
    @include utils.use-typography($typography);
    position: relative;
    display: inline-flex;
    flex-flow: row nowrap;
    gap: 1.6rem;
    align-items: center;
    cursor: pointer;

    &:is(.rc-disabled) {
      cursor: default;

      > * {
        color: map-get(map-get($colors, disabled), color);
      }
    }

    &.rc-label-before {
      flex-direction: row-reverse;
    }

    label {
      color: map-get($primary, color);
      cursor: inherit;
    }

    input {
      position: absolute;
      left: -9999px;
    }
  }
}
