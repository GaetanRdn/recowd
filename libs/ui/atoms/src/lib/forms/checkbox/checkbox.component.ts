import {
  ChangeDetectionStrategy,
  ChangeDetectorRef,
  Component,
  EventEmitter,
  forwardRef,
  HostListener,
  Input,
  NgModule,
  Output,
} from '@angular/core';
import { CommonModule } from '@angular/common';
import {
  BooleanInput,
  CoerceBoolean,
  Nullable,
  OnChangeFn,
  OnTouchedFn,
  TypedControlValueAccessor,
} from '@recowd/utility-types';
import { MaterialIconComponent } from '../../material-icon/material-icon.component';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { FormFieldElementDirective } from '../form-field/form-field-element.directive';

@Component({
  selector: 'rc-checkbox',
  host: {
    '[class.rc-label-before]': 'labelBefore',
    '[class.rc-disabled]': 'disabled',
  },
  templateUrl: './checkbox.component.html',
  providers: [
    {
      provide: NG_VALUE_ACCESSOR,
      useExisting: forwardRef(() => CheckboxComponent),
      multi: true,
    },
    {
      provide: FormFieldElementDirective,
      useExisting: forwardRef(() => CheckboxComponent),
    },
  ],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class CheckboxComponent<ValueType>
  implements
    TypedControlValueAccessor<Nullable<ValueType>>,
    FormFieldElementDirective
{
  static ngAcceptInputType_labelBefore: BooleanInput;
  static ngAcceptInputType_checked: BooleanInput;
  static ngAcceptInputType_disabled: BooleanInput;
  static ngAcceptInputType_required: BooleanInput;
  static ngAcceptInputType_hasError: BooleanInput;

  @Input() @CoerceBoolean() public labelBefore = false;

  @Input() @CoerceBoolean() public checked = false;

  @Input() @CoerceBoolean() public required = false;

  @Input() @CoerceBoolean() public hasError = false;

  @Input() public value!: ValueType;

  @Output() public readonly checkedChange: EventEmitter<
    CheckboxComponent<ValueType>
  > = new EventEmitter<CheckboxComponent<ValueType>>();

  constructor(private readonly _cdr: ChangeDetectorRef) {}

  private _disabled = false;

  get disabled(): boolean {
    return this._disabled;
  }

  @Input() @CoerceBoolean() set disabled(disabled: boolean) {
    this._disabled = disabled;
  }

  get focused(): boolean {
    return false;
  }

  @HostListener('click', ['$event'])
  public toggle(event?: Event): void {
    if (!this.disabled) {
      event?.stopPropagation();
      event?.stopImmediatePropagation();

      this.checked = !this.checked;
      this.onChange(this.checked ? this.value : null);
      this.onTouched();
      this.checkedChange.emit(this);
    }
  }

  public registerOnChange(fn: OnChangeFn<Nullable<ValueType>>): void {
    this.onChange = fn;
  }

  public registerOnTouched(fn: OnTouchedFn): void {
    this.onTouched = fn;
  }

  public writeValue(value: Nullable<ValueType>): void {
    this.checked = !!value;
  }

  public setDisabledState(isDisabled: boolean): void {
    this._disabled = isDisabled;
    this._cdr.markForCheck();
  }

  public onFormFieldClick(): void {
    this.toggle();
  }

  private onChange: OnChangeFn<Nullable<ValueType>> = () => {
    // default method
  };

  private onTouched: OnTouchedFn = () => {
    // default method
  };
}

@NgModule({
  declarations: [CheckboxComponent],
  exports: [CheckboxComponent],
  imports: [CommonModule, MaterialIconComponent],
})
export class CheckboxModule {}
