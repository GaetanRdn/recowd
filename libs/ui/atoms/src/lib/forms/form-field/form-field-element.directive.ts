import { Directive } from '@angular/core';

@Directive()
export abstract class FormFieldElementDirective {
  abstract get disabled(): boolean;

  abstract get focused(): boolean;

  abstract get required(): boolean;

  abstract get hasError(): boolean;

  public abstract onFormFieldClick(): void;
}
