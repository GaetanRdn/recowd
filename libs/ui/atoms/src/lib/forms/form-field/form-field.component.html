<ng-content select="rc-label"></ng-content>
<ng-content></ng-content>
