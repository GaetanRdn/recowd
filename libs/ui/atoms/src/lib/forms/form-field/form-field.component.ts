import {
  ChangeDetectionStrategy,
  ChangeDetectorRef,
  Component,
  ContentChild,
  HostListener,
  Input,
  NgModule,
} from '@angular/core';
import { LabelDirective } from '../label/label.directive';
import { FormFieldElementDirective } from './form-field-element.directive';
import { CoerceBoolean } from '@recowd/utility-types';

@Component({
  selector: 'rc-form-field',
  host: {
    '[class.rc-disabled]': 'disabled',
    '[class.rc-focused]': 'focused',
    '[class.rc-outlined]': 'outlined',
    '[class.rc-required]': 'required',
    '[class.rc-error]': 'hasError',
  },
  templateUrl: 'form-field.component.html',
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class FormFieldComponent {
  @ContentChild(FormFieldElementDirective, { static: true })
  public readonly formFieldElement?: FormFieldElementDirective;

  @Input() @CoerceBoolean() public outlined = false;

  get disabled(): boolean {
    return Boolean(this.formFieldElement?.disabled);
  }

  get focused(): boolean {
    return Boolean(this.formFieldElement?.focused);
  }

  get required(): boolean {
    return Boolean(this.formFieldElement?.required);
  }

  get hasError(): boolean {
    return Boolean(this.formFieldElement?.hasError);
  }

  constructor(private readonly _cdr: ChangeDetectorRef) {}

  @HostListener('click')
  private onClick(): void {
    this.formFieldElement?.onFormFieldClick();
    this._cdr.markForCheck();
  }
}

@NgModule({
  declarations: [FormFieldComponent],
  imports: [LabelDirective],
  exports: [FormFieldComponent, LabelDirective],
})
export class FormFieldModule {}
