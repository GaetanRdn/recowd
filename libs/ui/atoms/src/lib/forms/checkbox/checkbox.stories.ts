import { BADGE } from '@geometricpanda/storybook-addon-badges';
import { Meta, moduleMetadata, Story } from '@storybook/angular';
import { CheckboxComponent } from './checkbox.component';
import { action } from '@storybook/addon-actions';
import { ReactiveFormsModule, UntypedFormControl } from '@angular/forms';

export default {
  title: 'atoms/forms/checkbox',
  parameters: {
    badges: [BADGE.STABLE],
    jest: 'checkbox.component',
  },
  decorators: [
    moduleMetadata({
      imports: [CheckboxComponent, ReactiveFormsModule],
    }),
  ],
} as Meta<CheckboxComponent<number>>;

const template: Story<CheckboxComponent<number>> = (
  args: CheckboxComponent<number>
) => ({
  props: {
    labelBefore: args.labelBefore,
    checked: args.checked,
    disabled: args.disabled,
    log: action('checkedChange'),
  },
  template: `<rc-checkbox [labelBefore]="labelBefore" [checked]="checked" [value]="1" [disabled]="disabled" (checkedChange)="log($event)">Used</rc-checkbox>`,
});

export const Default = template.bind({});
Default.args = {
  labelBefore: false,
  checked: false,
  disabled: false,
};

const reactiveTemplate: Story<CheckboxComponent<number>> = (
  args: CheckboxComponent<number>
) => ({
  props: {
    labelBefore: args.labelBefore,
    control: new UntypedFormControl({
      value: args.checked ? 1 : null,
      disabled: args.disabled,
    }),
    log: action('checkedChange'),
  },
  template: `<rc-checkbox [labelBefore]="labelBefore" [formControl]="control" [value]="1" (checkedChange)="log($event)">Reactively Used</rc-checkbox>`,
});

export const Reactive = reactiveTemplate.bind({});
Reactive.args = {
  ...Default.args,
};
