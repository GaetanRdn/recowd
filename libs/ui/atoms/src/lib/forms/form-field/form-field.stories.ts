import { BADGE } from '@geometricpanda/storybook-addon-badges';
import { Args, Meta, moduleMetadata, Story } from '@storybook/angular';
import { FormFieldComponent, FormFieldModule } from './form-field.component';
import { InputDirective } from '../input/input.directive';
import { ReactiveFormsModule } from '@angular/forms';
import { CheckboxModule } from '../checkbox/checkbox.component';

export default {
  title: 'atoms/forms/form-field',
  component: FormFieldComponent,
  parameters: {
    badges: [BADGE.STABLE],
    jest: 'form-field.directive',
  },
  decorators: [
    moduleMetadata({
      imports: [
        FormFieldModule,
        InputDirective,
        CheckboxModule,
        ReactiveFormsModule,
      ],
    }),
  ],
} as Meta<FormFieldComponent>;

const templateWithInput: Story = (args: Args) => ({
  props: {
    outlined: args['outlined'],
    disabled: args['disabled'],
    required: args['required'],
  },
  template: `
    <rc-form-field [outlined]="outlined">
      <rc-label>Field label</rc-label>
      <input rcInput placeholder="Type something" [disabled]="disabled" [required]="required"/>
    </rc-form-field>`,
});

export const WithInput = templateWithInput.bind({});
WithInput.args = {
  outlined: false,
  disabled: false,
  required: false,
};

const templateWithCheckbox: Story = (args: Args) => ({
  props: {
    outlined: args['outlined'],
    disabled: args['disabled'],
    required: args['required'],
  },
  template: `
    <rc-form-field [outlined]="outlined">
      <rc-label>Field label</rc-label>
      <rc-checkbox [disabled]="disabled" [required]="required">Used</rc-checkbox>
    </rc-form-field>`,
});

export const WithCheckbox = templateWithCheckbox.bind({});
WithCheckbox.args = {
  outlined: false,
  disabled: false,
  required: false,
};
