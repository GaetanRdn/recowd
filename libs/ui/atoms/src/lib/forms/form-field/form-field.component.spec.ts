import { TestBed, waitForAsync } from '@angular/core/testing';
import { Component, CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { FormFieldComponent, FormFieldModule } from './form-field.component';
import { InputDirective } from '../input/input.directive';
import { TemplateLookup } from '@recowd/test/utils';
import { FormFieldElementDirective } from './form-field-element.directive';
import { CheckboxComponent } from '../checkbox/checkbox.component';

describe('FormFieldComponent', () => {
  beforeEach(waitForAsync(() =>
    TestBed.configureTestingModule({
      declarations: [HostWithInputComponent, HostWithCheckboxComponent],
      imports: [FormFieldModule, InputDirective, CheckboxComponent],
      schemas: [CUSTOM_ELEMENTS_SCHEMA],
    }).compileComponents()));

  describe('With input', () => {
    let templateLookup: TemplateLookup<HostWithInputComponent>;

    beforeEach(
      () => (templateLookup = new TemplateLookup(HostWithInputComponent))
    );

    it('should create With input', () => {
      templateLookup.detectChanges();
      expect(templateLookup.firstChildElement).toMatchSnapshot();
      expect(
        templateLookup.getComponent(FormFieldComponent).formFieldElement
      ).toEqual(templateLookup.getDirectiveOn('input', InputDirective));
    });

    it('should create With disabled input', () => {
      jest
        .spyOn(
          templateLookup.getComponent(FormFieldComponent)
            .formFieldElement as FormFieldElementDirective,
          'disabled',
          'get'
        )
        .mockReturnValue(true);
      templateLookup.detectChanges();
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });

    it('should known when input is focused', () => {
      templateLookup.query('input').focus();
      templateLookup.detectChanges();
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });

    it('should set the focus on input when click', () => {
      templateLookup.query('rc-form-field').click();
      templateLookup.detectChanges();
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });

    it('should be outlined', () => {
      templateLookup.hostComponent.outlined = true;
      templateLookup.detectChanges();
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });

    it('should be required', () => {
      jest
        .spyOn(
          templateLookup.getComponent(FormFieldComponent)
            .formFieldElement as FormFieldElementDirective,
          'required',
          'get'
        )
        .mockReturnValue(true);
      templateLookup.detectChanges();
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });

    it('should be in error', () => {
      jest
        .spyOn(
          templateLookup.getComponent(FormFieldComponent)
            .formFieldElement as FormFieldElementDirective,
          'hasError',
          'get'
        )
        .mockReturnValue(true);
      templateLookup.detectChanges();

      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });
  });

  describe('With checkbox', () => {
    let templateLookup: TemplateLookup<HostWithCheckboxComponent>;

    beforeEach(
      () => (templateLookup = new TemplateLookup(HostWithCheckboxComponent))
    );

    it('should create With disabled input', () => {
      jest
        .spyOn(
          templateLookup.getComponent(FormFieldComponent)
            .formFieldElement as FormFieldElementDirective,
          'disabled',
          'get'
        )
        .mockReturnValue(true);
      templateLookup.detectChanges();
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });

    it('should known when input is focused', () => {
      templateLookup.query('input').focus();
      templateLookup.detectChanges();
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });

    it('should set the focus on input when click', () => {
      templateLookup.query('rc-form-field').click();
      templateLookup.detectChanges();
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });

    it('should be outlined', () => {
      templateLookup.hostComponent.outlined = true;
      templateLookup.detectChanges();
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });

    it('should be required', () => {
      jest
        .spyOn(
          templateLookup.getComponent(FormFieldComponent)
            .formFieldElement as FormFieldElementDirective,
          'required',
          'get'
        )
        .mockReturnValue(true);
      templateLookup.detectChanges();
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });

    it('should be in error', () => {
      jest
        .spyOn(
          templateLookup.getComponent(FormFieldComponent)
            .formFieldElement as FormFieldElementDirective,
          'hasError',
          'get'
        )
        .mockReturnValue(true);
      templateLookup.detectChanges();

      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });
  });
});

@Component({
  template: ` <rc-form-field [outlined]="outlined">
    <rc-label>Field label</rc-label>
    <input rcInput />
  </rc-form-field>`,
})
class HostWithInputComponent {
  public outlined = false;
}

@Component({
  template: ` <rc-form-field [outlined]="outlined">
    <rc-label>Field label</rc-label>
    <rc-checkbox [value]="1"></rc-checkbox>
  </rc-form-field>`,
})
class HostWithCheckboxComponent {
  public outlined = false;
}
