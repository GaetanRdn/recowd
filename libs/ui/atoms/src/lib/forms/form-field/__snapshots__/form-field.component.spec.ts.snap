// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormFieldComponent should be in error 1`] = `
<rc-form-field
  class="rc-error"
>
  <rc-label>
    Field label
  </rc-label>
  <input
    class="rc-input"
    rcinput=""
  />
</rc-form-field>
`;

exports[`FormFieldComponent should be outlined 1`] = `
<rc-form-field
  class="rc-outlined"
>
  <rc-label>
    Field label
  </rc-label>
  <input
    class="rc-input"
    rcinput=""
  />
</rc-form-field>
`;

exports[`FormFieldComponent should be required 1`] = `
<rc-form-field
  class="rc-required"
>
  <rc-label>
    Field label
  </rc-label>
  <input
    class="rc-input"
    rcinput=""
    required=""
  />
</rc-form-field>
`;

exports[`FormFieldComponent should create With disabled input 1`] = `
<rc-form-field
  class="rc-disabled"
>
  <rc-label>
    Field label
  </rc-label>
  <input
    class="rc-input"
    disabled=""
    rcinput=""
  />
</rc-form-field>
`;

exports[`FormFieldComponent should create With input 1`] = `
<rc-form-field>
  <rc-label>
    Field label
  </rc-label>
  <input
    class="rc-input"
    rcinput=""
  />
</rc-form-field>
`;

exports[`FormFieldComponent should known when input is focused 1`] = `
<rc-form-field
  class="rc-focused"
>
  <rc-label>
    Field label
  </rc-label>
  <input
    class="rc-input"
    rcinput=""
  />
</rc-form-field>
`;

exports[`FormFieldComponent should set the focus on input when click 1`] = `
<rc-form-field
  class="rc-focused"
>
  <rc-label>
    Field label
  </rc-label>
  <input
    class="rc-input"
    rcinput=""
  />
</rc-form-field>
`;
